runtime/dart_isolate.cc

```diff
@@ -280,8 +284,130 @@ std::weak_ptr<DartIsolate> DartIsolate::CreateRootIsolate(
+  // std::cout << "RuntimeController: " << (void*)runtime << std::endl;
+
+  // TODO: This doesn't work for platform isolates spawned from other child
+  //     isolates, since their TaskRunners are null. Need a better way of
+  //     getting them. Maybe child isolates should have non-null TaskRunners?
+  const TaskRunners& task_runners = GetTaskRunners();
+  /*((RuntimeController*)platform_configuration()->client())
+      ->GetRootIsolate()
+      .lock()
+      ->GetTaskRunners();*/
+  auto isolate_group_data =
+      std::make_unique<std::shared_ptr<DartIsolateGroupData>>(
+          std::shared_ptr<DartIsolateGroupData>(
+              *static_cast<std::shared_ptr<DartIsolateGroupData>*>(
+                  Dart_IsolateGroupData(parent_isolate))));
+
+  std::cout << "Task runner: "
+            << (void*)task_runners.GetPlatformTaskRunner().get() << std::endl;
+
+  UIDartState::Context context(task_runners);
+  context.advisory_script_uri = (*isolate_group_data)->GetAdvisoryScriptURI();
+  context.advisory_script_entrypoint =
+      (*isolate_group_data)->GetAdvisoryScriptEntrypoint();
+  auto isolate_data = std::make_unique<std::shared_ptr<DartIsolate>>(
+      std::shared_ptr<DartIsolate>(
+          new DartIsolate((*isolate_group_data)->GetSettings(),  // settings
+                          false,       // is_root_isolate
+                          true,        // is_platform_isolate
+                          context)));  // context
+  (*isolate_data)->platform_isolate_manager_ = platform_isolate_manager;
+
+  IsolateMaker isolate_maker =
+      [parent_isolate, debug_name](
+          std::shared_ptr<DartIsolateGroupData>* isolate_group_data,
+          std::shared_ptr<DartIsolate>* isolate_data, Dart_IsolateFlags* flags,
+          char** error) {
+        std::cout << "           PlatformIsolate data: " << (void*)isolate_data
+                  << std::endl;
+        // TODO: Are those the right shutdown_callback and cleanup_callback?
+        return Dart_CreateIsolateInGroup(
+            /*group_member=*/parent_isolate,
+            /*name=*/debug_name.c_str(),
+            /*shutdown_callback=*/
+            reinterpret_cast<Dart_IsolateShutdownCallback>(
+                DartIsolate::SpawnIsolateShutdownCallback),
+            /*cleanup_callback=*/
+            reinterpret_cast<Dart_IsolateCleanupCallback>(
+                DartIsolateCleanupCallback),
+            /*child_isolate_data=*/isolate_data,
+            /*error=*/error);
+        // TODO: Passing a non-null isolate_data causes platform isolate to
+        //     never exit.
+      };
+  DartErrorString error;
+  Dart_Isolate new_isolate = CreateDartIsolateGroup(
+      std::move(isolate_group_data), std::move(isolate_data), nullptr,
+      error.error(), isolate_maker);
+
+  if (error) {
+    FML_LOG(ERROR) << "CreatePlatformIsolate failed: " << error.str();
+    return false;
+  }
+
+  platform_isolate_manager->RegisterPlatformIsolate(new_isolate);
```

Comment:
Is there a race between this and the platform thread potentially calling `ShutdownPlatformIsolates`?

`RegisterPlatformIsolate` has an assertion that assumes that `is_shutdown_` is false.  But IIUC it's possible for shell shutdown to call `ShutdownPlatformIsolates` on the platform thread between the `platform_isolate_manager->IsShutdown()` call above and this point.

Reply:
Oh right. I've tightened up this sequence. I think it should be fixed now.

`RegisterPlatformIsolate` now double checks `is_shutdown_`, and returns false if it is shutdown. In that case the caller stops setting up the isolate and does nothing else (ie doesn't explicitly shut it down). This is fine because this case means that the engine is being shutdown, so the isolate is about to be shutdown anyway.

`RemovePlatformIsolate` and `ShutdownPlatformIsolates` are only ever called on the platform thread, so we don't have the same race condition between them.

I've added comments about all this stuff too.